Send Custom Profile Field Values in the /json/users Endpoint

Hypro999 opened the pull request:
https://chat.zulip.org/#narrow/stream/3-backend/topic/Custom.20profile.20fields.20in.20the.20API

Hypro999 commented:
I haven't added the change for profile yet, because for some reason defining `get_profile_backend` as

```python3
def get_profile_backend(request: HttpRequest, user_profile: UserProfile,
                        include_custom_profile_fields: bool=REQ(validator=check_bool,
                                                                default=False)
                        ) -> HttpResponse:
```

causes `include_custom_profile_fields` to be an instance of `_REQ` instead of a `bool`. I'm pretty sure that I'm making a really small mistake that's leading to this behavior but I couldn't figure it out today.

The test I was using for this is:
```python3
    def test_get_custom_profile_fields_from_api_for_self(self) -> None:
        self.login(self.example_email("iago"))

        # url = "/json/users/me"
        # response = self.client_get(url)
        # self.assertEqual(response.status_code, 200)
        # data = response.json()

        url = "/json/users/me?include_custom_profile_fields=true"
        response = self.client_get(url)
        self.assertEqual(response.status_code, 200)
        data = response.json()
        # self.assertTrue("profile_data" in data.keys())
        import json
        print(json.dumps(data, indent=4))
        with self.assertRaises(KeyError):
            data["profile_data"]
```

timabbott commented:
The issue there is that you need to use `@has_request_variables` on an endpoint for REQ to be processed.

timabbott commented on zerver/tests/test_bots.py:
```diff
@@ -146,7 +146,7 @@ def test_json_users_with_bots(self) -> None:
         self.assert_json_success(users_result)
 
-        self.assert_length(queries, 3)
+        self.assert_length(queries, 9)
```
This change is wrong; the fact that this number is going from 3 to 9 means you have database queries in a loop.

A quick print-debug suggests it's UserProfile fetches.

Hypro999 replied:
4 of them if I remember. I was concerned about this too.

timabbott commented on zerver/lib/events.py:
```diff
@@ -102,13 +108,24 @@ def user_data(row: Dict[str, Any]) -> Dict[str, Any]:
+                    bot_owner = UserProfile.objects.get(id=bot_owner_id)
+                    result['bot_owner'] = bot_owner.email
+                    if include_delivery_emails:
+                        result['bot_owner_delivery_email'] = bot_owner.delivery_email
```
I guess this is where you're doing that query.  I think ultimately the right fix for this is that we shouldn't be sending bot owners as an email address; we should be sending bot owner user IDs instead.  

I just checked and the mobile apps don't use the `bot_owner` field, so the right solution is to just do a prep commit series that changes `bot_owner` to be an integer ID for the bot's owner; that's a cleaner API anyway in several ways, and lets us avoid adding this mess.

Hypro999 replied:
Sounds great! I was hesitant to make such a change initially because changing the API means affecting multiple people (the mobile apps as well as other (non-Zulip) developers who are using the API, like people who make bots). But I'll go ahead and make this change esp. since the bot owner's email can still be obtained from the data that this endpoint is returning without needing to make an extra request, if we at least send the id of the owner.

I'll make the change. :+1:

timabbott replied:
Yeah, the key detail here is that none of the other apps look at the `bot_owner` field.

timabbott commented on zerver/lib/events.py:
```diff
@@ -74,6 +72,14 @@ def get_raw_user_data(realm: Realm, client_gravatar: bool) -> Dict[int, Dict[str
+                      include_custom_profile_fields: bool=True,
+                      include_delivery_emails: bool=False) -> Dict[int, Dict[str, str]]:
+    user_dicts = get_realm_user_dicts(realm.id)
+    profiles_by_user_id = get_custom_profile_field_values(realm.id)
```
It seems like we could skip this query if `include_custom_profile_fields` is `False`.

Hypro999 replied:
Yes, good catch! Thanks for pointing it out!

Hypro999 commented:
Force pushing to deal with the Travis flake.

Hypro999 commented:
@timabbott I think that this is ready for review. I've also added a commit for sending custom profile fields with the /users/me (profile) endpoint.
So the commits are:
1. Extract the custom profile field values gathering code.
2. Modify `get_raw_user_data` and apply it to the `/users` endpoint.
3. Update the `/users/me` endpoint.

timabbott commented:
Thanks @Hypro999!   I've merged the first couple commits, after some small changes to comments and API docs.  Pushed the last commit back here.

For the third commit, there are some issues:
* I think for `get_profile_backend`, we can just include those fields unconditionally; for a single user, it's inexpensive to provide; it's only for bulk queries where we'd want to filter.
* It duplicates a lot of code from `zerver/lib/events.py` for formatting; I think with a bit more refactoring we could avoid that.
* The commit message is copied from the second commit :)

Also a few follow-up tasks:
* We should also add a follow-up commit removing that `for_api` parameter; I think we should never want to include a `bot_type` parameter in the response to `GET /users`, since non-bot users have no bot type.  Or am I missing a case where that code runs for a bot user?
* Can you check that we don't have a response fixture we need to update in the API docs for both of these endpoints?  
* Can you look at `python-zulip-api` to see if we should add the custom profile fields parameter as an argument in the Python function?  Not sure if it supports passing an arguments dict or not.

Hypro999 commented:
> I think for get_profile_backend, we can just include those fields unconditionally; for a single user, it's inexpensive to provide; it's only for bulk queries where we'd want to filter.

Sounds good; done.

> The commit message is copied from the second commit.

Oops, I really don't know how that happened though :sweat_smile:

Hypro999 commented:
> Or am I missing a case where that code runs for a bot user?

Nope, you're not missing anything and I completely agree. As you've correctly pointed out in the comment you've added to commit `1946692f9`, the only reason for sending the `bot_type` unconditionally is because endpoint was previously sending them unconditionally.

Hypro999 commented:
> Can you check that we don't have a response fixture we need to update in the API docs for both of these endpoints?

Yup, I needed to update the documentation for these endpoints.

Hypro999 commented:
- [x] I think for get_profile_backend, we can just include those fields unconditionally; for a single user, it's inexpensive to provide; it's only for bulk queries where we'd want to filter. -> Sure thing; done.

- [x] The commit message is copied from the second commit :) -> A mistake; fixed.

- [x] We should also add a follow-up commit removing that for_api parameter; I think we should never want to include a bot_type parameter in the response to GET /users, since non-bot users have no bot type. Or am I missing a case where that code runs for a bot user? -> Originally this was done to maintain compatibility with what was being done before. I removed the for_api parameter and am now only sending bot_type for bots.

- [x] Can you check that we don't have a response fixture we need to update in the API docs for both of these endpoints? -> Updated response fixtures.

- [ ] Can you look at python-zulip-api to see if we should add the custom profile fields parameter as an argument in the Python function? Not sure if it supports passing an arguments dict or not. -> will do.

Hypro999 commented on zerver/openapi/zulip.yaml:
```diff
@@ -1232,6 +1266,10 @@ paths:
+                    profile_data:
+                      type: object
+                      description: A object mapping the id of each custom profile field id to the
+                        value and rendered_value for it (if it exists) for the requesting user.
```
@timabbott I definitely think that this could be worded better. But I also think that you already have a nice way in mind for this :slightly_smiling_face:

Hypro999 commented:
@timabbott this is ready for review now :slightly_smiling_face:

timabbott commented:
Nice, merged after tweaking those docs, thanks @Hypro999!

This was a great series of cleanups, I'm very glad you made them happen.  Ping me on Zulip if you need help picking a next project.
